libvmaf/src/feature/cambi.c

```diff
@@ -912,36 +912,35 @@ static FORCE_INLINE inline double weight_scores_per_scale(double *scores_per_sca
     return score / normalization;
 }
 
-static void write_uint16(FILE *file, uint16_t v) {
-    fwrite((void*)(&v), sizeof(v), 1, file);
-}
-
-
-static int dump_c_values(const float *c_values, int width, int height, int scale, 
-                          int window_size, const uint16_t num_diffs) {
-    int max_diff_weight = g_diffs_weights[0];
+static int dump_c_values(FILE *heatmaps_files[], const float *c_values, int width, int height, int scale, 
+                         int window_size, const uint16_t num_diffs, const int *diff_weights, int frame) {
+    int max_diff_weight = diff_weights[0];
     for (int i = 0; i < num_diffs; i++) {
-        if (g_diffs_weights[i] > max_diff_weight) {
-            max_diff_weight = g_diffs_weights[i];
+        if (diff_weights[i] > max_diff_weight) {
+            max_diff_weight = diff_weights[i];
         }
     }
     int max_c_value = max_diff_weight * window_size * window_size / 4;
     int max_16bit_value = (1 << 16) - 1;
     double scaling_value = (double)max_16bit_value / max_c_value;
-    FILE *file = g_heatmaps_paths[scale];
+    FILE *file = heatmaps_files[scale];
+    uint16_t *to_write = malloc(width * sizeof(uint16_t));
     for (int i = 0; i < height; i++) {
         for (int j = 0; j < width; j++) {
-            uint16_t val = (uint16_t)(scaling_value * c_values[i * width + j]);
-            write_uint16(file, val);
+            to_write[j] = (uint16_t)(scaling_value * c_values[i * width + j]);
         }
+        ptrdiff_t offset = (frame * height + i) * width * sizeof(uint16_t);
+        fseek(file, offset, SEEK_SET);
```

Comment:
So I've tried this `fseek` idea and it seems to work, with `--threads 8` it has generated the correct heatmap in the ~5 times I've tried it so (at least on my machine/os) it is thread-safe. @kylophone any concern?

Is it a problem to have the malloc in line 927? My concern with doing it in `init` is that there are already way too many parameters to these functions and it's quite messy

Reply:
> So I've tried this `fseek` idea and it seems to work, with `--threads 8` it has generated the correct heatmap in the ~5 times I've tried it so (at least on my machine/os) it is thread-safe. @kylophone any concern?

Should be OK. No file pointers (FILE*) are shared between threads, so I don't think there will be any issues. Performance is probably not going to be great, but better than a jumbled mess. 

> Is it a problem to have the malloc in line 927? My concern with doing it in `init` is that there are already way too many parameters to these functions and it's quite messy

It's probably ok, but it's an unchecked malloc right now.